src/debug.c

```diff
@@ -2387,74 +2471,108 @@ void debugDelay(int usec) {
     if (usec) usleep(usec);
 }
 
+#ifdef HAVE_BACKTRACE
 #ifdef __linux__
-#include <sys/syscall.h>
-
-#define THREADS_NUMBER 2
-static const size_t buff_len = 256;
-/* tids of additional threads + main thread */
-static pid_t test_tids[THREADS_NUMBER + 1];
-static volatile int wait_for_signal = 1;
-static redisAtomic size_t g_done = 0;
 
+/* =========================== Stacktrace Utils ============================ */
+#define TIDS_INITIAL_SIZE 50
+
+/** If it doesn't block and doesn't ignore, return 1 (the thread will handle the signal) 
+ * If thread tid blocks or ignores sig_num returns 0 (thread is not ready to catch the signal).
+ * also returns 0 if it failed to open the thread's status file. **/
+static int is_thread_ready_to_signal(pid_t pid, pid_t tid, int sig_num) {
+    /* open the threads status file */
+    char buff[MAX_BUFF_LENGTH];
+    snprintf(buff, MAX_BUFF_LENGTH, "/proc/%d/task/%d/status", pid, tid);
+    FILE *thread_status_file = fopen(buff, "r");
+    if (thread_status_file == NULL) {
+        return 0;
+    }
 
-static void *thread_do(void *arg) {
-    size_t thread_id = (size_t)arg;
-    test_tids[thread_id] = syscall(SYS_gettid);
+    int ret = 1;
 
-    atomicIncr(g_done, 1);
-    while(wait_for_signal) {}
+    size_t field_name_len = strlen("SigBlk:");
+    while (fgets(buff, MAX_BUFF_LENGTH, thread_status_file)) {
+        /* iterate the file until we reach SigBlk field line*/
+        if (!strncmp(buff, "SigBlk:", field_name_len)) {
+            /* check if the signal is blocked by this thread */
+            unsigned long sig_block_mask = strtoul(buff + field_name_len, NULL, 16);
+            if(sig_block_mask & sig_num) {
+                ret = 0;
+            } else {
+                /* get SigIgn field line and check if this thread ignores sig_num*/
+                if (fgets(buff, MAX_BUFF_LENGTH, thread_status_file)) {/* Avoid warning. */};
```

Comment:
It is not explicitly guaranteed in the Linux documentation, but this is the example format, and in addition, they are grouped together 
https://man7.org/linux/man-pages/man5/proc.5.html
![Untitled](https://github.com/redis/redis/assets/109056284/f3f12d57-d269-4e76-865f-8bb63f332da7)
![image](https://github.com/redis/redis/assets/109056284/7f5d0379-d80f-45d3-b3fd-5be9a8ba0600)
Anyway, it is safer to not rely on the file order. I'll change it.
Also I'll add a warning msg in case one of the fields is not found and the test will check that this warning doesn't appear.